tests/test_annotations.py

```diff
@@ -0,0 +1,199 @@
+        field for each file in the returned tarball. See tarfile documentation
+        for more details.
+    """
+    b = "21\t.\tbiological_region\t6111184\t6111187\t0.999\t+\t.\tlogic_name=eponine\n"
+    b += b
+    b += b
+
+    with tempfile.TemporaryDirectory() as map_dir:
+        with open(os.path.join(map_dir, "test.gff"), "w") as f:
+            print("##gff-version 3", file=f)
+            print("##sequence-region   21 1 46709983", file=f)
+            print(b, file=f)
+
+        if custom_file_f is not None:
+            # Do arbitrary things under map_dir.
+            custom_file_f(map_dir)
+
+        # For the zipfile to be in the right format, we must be in the right directory.
+        with open(os.path.join(map_dir, "test.gff"), 'rb') as f_in:
+            # Now zip this
+            with gzip.open("test.gff.gz", mode="wb") as gz_file:
+                shutil.copyfileobj(f_in, gz_file)
+        # Read back the tarball
+        an = gzip.open("test.gff.gz", mode='rb')
+    return an
+
+
+class TestGTF(tests.CacheWritingTest):
+    """
+    Tests for the basic functionality of the genetic map class.
+    """
+
+    def test_cache_dirs(self):
+        gm = GTFTestClass()
+        cache_dir = stdpopsim.get_cache_dir() / "annotations"
+        self.assertEqual(gm.cache_dir, cache_dir)
+        self.assertEqual(gm.species_cache_dir, gm.cache_dir / gm.species.id)
+
+    def test_str(self):
+        gm = GTFTestClass()
+        self.assertGreater(len(str(gm)), 0)
+
+    def test_is_cached(self):
+        gm = GTFTestClass()
+        os.makedirs(gm.species_cache_dir, exist_ok=True)
+        self.assertTrue(gm.is_cached())
+        shutil.rmtree(gm.cache_dir)
+        self.assertFalse(gm.is_cached())
+
+
+class TestGTFDownload(tests.CacheWritingTest):
+    """
+    Tests downloading code for the genetic maps.
+    """
+
+    def test_correct_url(self):
+        gm = GTFTestClass()
+        with mock.patch("urllib.request.urlretrieve", autospec=True) as mocked_get:
+            # The destination file will be missing.
+            print(gm.url)
+            with self.assertRaises(ValueError):
+                gm.download()
+        mocked_get.assert_called_once_with(gm.url, filename=unittest.mock.ANY)
+
+    def test_download_over_cache(self):
+        for gm in stdpopsim.all_annotations():
+            gm.download()
+            self.assertTrue(gm.is_cached())
+            gm.download()
+            self.assertTrue(gm.is_cached())
+
+    def test_multiple_threads_downloading(self):
+        gm = next(stdpopsim.all_annotations())
+        gm.download()
+        saved = gm.is_cached
+        try:
+            # Trick the download code into thinking there's several happening
+            # concurrently
+            gm.is_cached = lambda: False
+            with self.assertWarns(Warning):
+                gm.download()
+        finally:
+            gm.is_cached = saved
+
+
+class TestAllGeneticMaps(tests.CacheReadingTest):
```

Comment:
'TestAllGeneticMaps'... It looks like this is not related to annotations or the GTF class?